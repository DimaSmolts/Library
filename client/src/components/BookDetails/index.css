.book-detail-container {
  background: var(--gold);
  padding: 10px;
  margin: 25px 0;
  width: 100%;
  display: grid;
  grid-template-columns: 4fr 6fr;
  grid-template-rows: 10% 15% 10% 35% 10% auto;
  grid-gap: 1em;
  box-shadow: 5px 5px 15px;
  box-sizing: border-box;
  border-radius: 40px;
  border: 2px solid var(--dark-blue);
}

.book-detail-image {
  grid-row: 1 / 6;
  grid-column: 1 / 1;
}

.book-detail-image img {
  width: 100%;
  border-radius: 35px;
  border: 2px solid var(--dark-blue);
}

.book-detail-rate {
  grid-row: 6 / 7;
  grid-column: 1 / 1;

  display: grid;
  grid-template-columns: repeat(10, auto);
  justify-content: space-between;
}

.book-detail-rate img {
  width: 80%;
  justify-self: center;
}

.book-detail-author {
  font-size: 2em;
  align-self: center;
}

.book-detail-name {
  font-size: 3em;
  align-self: center;
}

.book-detail-purchase {
  justify-self: end;
  align-self: center;
  grid-row: 5 / 7;
}

.book-detail-price {
  font-size: 2em;
  margin-right: 15px;
}

.book-detail-purchase-button {
  height: 50px;
  font-size: 1em;
  font-weight: bold;
  background: var(--gold);
  border-radius: 40px;
  border: 4px solid var(--dark-blue);
  vertical-align: bottom;
}

.book-detail-purchase-button:hover {
  background: var(--dark-gold);
  cursor: pointer;
}

.book-detail-description {
  text-align: justify;
  text-indent: 2em;
}

.book-detail-genre {
  border-radius: 15px;
  border: 2px solid var(--dark-blue);
  padding: 0 6px;
  margin-right: 10px;
}

@media (max-width: 576px) {
  .book-detail-container {
    grid-template-rows: auto;
    grid-template-columns: 1fr;
  }
  .book-detail-image {
    grid-row: 1 / 2;
  }
  .book-detail-rate {
    grid-row: 2 / 3;
  }
  .book-detail-purchase {
    grid-row: 7/ 8;
  }
}
