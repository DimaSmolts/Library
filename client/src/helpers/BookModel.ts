export class Book {
  id: number;
  name: string;
  author: string;
  genres: Array<string>;
  description: string;
  rate: number;
  price: number;
  image: string;
}
