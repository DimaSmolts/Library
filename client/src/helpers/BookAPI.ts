import axios from 'axios';
import { Book } from './BookModel'

const baseURL = process.env.REACT_APP_API_URL || 'http://localhost:3000';

const instance = axios.create({
  baseURL: baseURL + '/api',
})

export const bookAPI = {
  async getBooks() {
    return await instance.get<Array<Book>>('/books');
  },
  async getBook(id: number) {
    return await instance.get<Book>('/books/' + id);
  }
}
