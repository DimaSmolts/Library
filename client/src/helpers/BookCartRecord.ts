import { Book } from './BookModel'

export interface BookCartRecord {
  book: Book,
  quantity: number,
}
