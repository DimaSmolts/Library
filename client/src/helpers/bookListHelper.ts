import { Book } from './BookModel'

export function filterBookList(books: Array<Book>, query: string): Array<Book> {
  return books.filter(book => {
    return book.name
      .trim()
      .toLowerCase()
      .includes(query.trim().toLowerCase())
  })
}
