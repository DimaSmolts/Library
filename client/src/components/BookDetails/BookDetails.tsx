import React from 'react'
import { Book } from '../../helpers/BookModel'
import './index.css'

const starSolid = require('../../assets/images/star-solid.svg')
const starOutline = require('../../assets/images/star-outline.svg')

interface BookDetailsProps {
  book: Book,
  quantityOfAddedItems: number,
  addBookToCart: (book: Book) => void,
  removeBookFromCart: (book: Book) => void,
}

const BookDetails: React.SFC<BookDetailsProps> = (props: BookDetailsProps) => {
  return <div className='book-detail-container'>
    <div className='book-detail-image'>
      <img src={props.book.image} />
    </div>
    <div className='book-detail-rate'>
      {
        (() => {
          let stars = Array<JSX.Element>()
          for (let i = 0; i < 10; i++) {
            if (i < props.book.rate) {
              stars.push(<img src={starSolid} />)
            } else {
              stars.push(<img src={starOutline} />)
            }
          }
          return stars
        })()
      }
    </div>
    <h3 className='book-detail-author'>
      {props.book.author}
    </h3>
    <h2 className='book-detail-name'>
      {props.book.name}
    </h2>
    <div className='book-detail-genres'>
      {
        props.book.genres
          .map(genre => {
            return <span className='book-detail-genre'>{genre}</span>
          })
      }
    </div>
    <div className='book-detail-description'>
      {props.book.description}
    </div>
    <div className='book-detail-purchase'>
      <span className='book-detail-price'>{props.book.price}Â£</span>
      <span>In cart: {props.quantityOfAddedItems}</span>
      <button
        onClick={() => props.addBookToCart(props.book)}
        className='book-detail-purchase-button'>add</button>
      <button
        onClick={() => props.removeBookFromCart(props.book)}
        disabled={props.quantityOfAddedItems === 0 ? true : false}
        className='book-detail-purchase-button'>remove</button>
    </div>
  </div>
}

export default BookDetails
