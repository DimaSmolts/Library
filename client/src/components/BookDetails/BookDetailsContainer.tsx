import React from 'react';
import { BookNavParams } from '../BookPage/BookPage'
import BookDetails from './BookDetails'
import Message, { MessageTypes, MessageTextSizes } from '../Message/Message'
import { getBookThunk, cleanUpThunk } from '../../store/thunks/bookThunks'
import { connect, ConnectedProps } from 'react-redux';
import { AppState } from '../../store/rootReducer'


class BookDetailsContainer extends React.Component<BookNavParams & PropsFromRedux> {
  componentDidMount() {
    this.props.getBook(+this.props.bookId);
  }

  componentWillUnmount() {
    this.props.cleanUp();
  }

  render() {
    if (this.props.error)
      throw new Error(this.props.error.toString());

    return <>
      {
        this.props.isFetched ?
          <BookDetails book={this.props.book} />
          :
          <Message type={MessageTypes.msgInfo} size={MessageTextSizes.msgBig} text='Loading...' />
      }
    </>
  }
}

const mapState = (state: AppState) => ({
  book: state.bookReducer.book,
  isFetched: state.bookReducer.isFetched,
  error: state.bookReducer.error
})


const mapDispatch = {
  getBook: (id: number) => getBookThunk(id),
  cleanUp: () => cleanUpThunk(),
}

const connector = connect(mapState, mapDispatch);
type PropsFromRedux = ConnectedProps<typeof connector>

export default connector(BookDetailsContainer)
