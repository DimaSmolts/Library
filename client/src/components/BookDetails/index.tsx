import React from 'react'
import { BookNavParams } from '../BookPage/BookPage'
import BookDetails from './BookDetails'
import Message, { MessageTypes, MessageTextSizes } from '../Message/Message'
import { getBookThunk, cleanUpThunk } from '../../store/thunks/bookThunks'
import { connect, ConnectedProps } from 'react-redux'
import { AppState } from '../../store/rootReducer'
import { Book } from '../../helpers/BookModel'
import { addBookToCartThunk, removeBookFromCartThunk } from '../../store/thunks/cartThunks'

class BookDetailsContainer extends React.Component<BookNavParams & PropsFromRedux> {
  componentDidMount() {
    this.props.getBook(+this.props.bookId)
  }

  componentWillUnmount() {
    this.props.cleanUp()
  }

  render() {
    if (this.props.error)
      throw new Error(this.props.error.toString())
    if (this.props.notFound)
      return <Message type={MessageTypes.msgNotFound} size={MessageTextSizes.msgBig} text='Not Found' />

    return <>
      {
        this.props.isFetched ?
          <BookDetails
            book={this.props.book}
            quantityOfAddedItems={this.props.quantityOfAddedItems}
            addBookToCart={this.props.addBookToCart}
            removeBookFromCart={this.props.removeBookFromCart} />
          :
          <Message type={MessageTypes.msgInfo} size={MessageTextSizes.msgBig} text='Loading...' />
      }
    </>
  }
}

const mapState = (state: AppState) => ({
  book: state.bookReducer.book,
  isFetched: state.bookReducer.isFetched,
  error: state.bookReducer.error,
  notFound: state.bookReducer.notFound,
  quantityOfAddedItems: state.cartReducer.bookRecords.find(br => br.book.id === state.bookReducer.book ?.id) ?.quantity || 0,
})

const mapDispatch = {
  getBook: (id: number) => getBookThunk(id),
  cleanUp: () => cleanUpThunk(),
  addBookToCart: (book: Book) => addBookToCartThunk(book),
  removeBookFromCart: (book: Book) => removeBookFromCartThunk(book),
}

const connector = connect(mapState, mapDispatch)
type PropsFromRedux = ConnectedProps<typeof connector>

export default connector(BookDetailsContainer)
