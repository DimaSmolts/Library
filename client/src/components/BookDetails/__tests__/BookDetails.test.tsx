import React from 'react'
import { shallow, mount } from 'enzyme'

import BookDetails from '../BookDetails'
import { Book } from '../../../helpers/BookModel'

const book = {
  id: 1,
  name: 'Name',
  author: 'Author',
  rate: 8,
  image: 'image',
  genres: ['first', 'second'],
  description: 'description',
  price: 20
} as Book
const defaultCount = 0
const notZeroCount = 2
const addBookToCart = jest.fn()
const removeBookFromCart = jest.fn()

test('snapshot: <BookDetails /> renders correctly', () => {
  const wrapper = shallow(
    <BookDetails
      book={book}
      quantityOfAddedItems={defaultCount}
      addBookToCart={addBookToCart}
      removeBookFromCart={removeBookFromCart} />)

  expect(wrapper).toMatchSnapshot()
})

it('should add book', () => {
  const wrapper = shallow(
    <BookDetails
      book={book}
      quantityOfAddedItems={defaultCount}
      addBookToCart={addBookToCart}
      removeBookFromCart={removeBookFromCart} />)

  wrapper.find('.add-book-button').simulate('click')
  expect(addBookToCart).toBeCalled()
})

it('should remove book', () => {
  const wrapper = shallow(
    <BookDetails
      book={book}
      quantityOfAddedItems={defaultCount}
      addBookToCart={addBookToCart}
      removeBookFromCart={removeBookFromCart} />)

  wrapper.find('.remove-book-button').simulate('click')
  expect(removeBookFromCart).toBeCalled()
})

it('should have remove button disabled', () => {
  const wrapper = shallow(
    <BookDetails
      book={book}
      quantityOfAddedItems={defaultCount}
      addBookToCart={addBookToCart}
      removeBookFromCart={removeBookFromCart} />)

  expect(wrapper.find('.remove-book-button').prop('disabled')).toBeTruthy();
})

it('should not have remove button disabled', () => {
  const wrapper = shallow(
    <BookDetails
      book={book}
      quantityOfAddedItems={notZeroCount}
      addBookToCart={addBookToCart}
      removeBookFromCart={removeBookFromCart} />)

  expect(wrapper.find('.remove-book-button').prop('disabled')).toBeFalsy();
})
