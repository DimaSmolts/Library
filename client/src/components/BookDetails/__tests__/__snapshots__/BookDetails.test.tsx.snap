// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot: <BookDetails /> renders correctly 1`] = `
<div
  className="book-detail-container"
>
  <div
    className="book-detail-image"
  >
    <img
      src="image"
    />
  </div>
  <div
    className="book-detail-rate"
  >
    <img
      src={Object {}}
    />
    <img
      src={Object {}}
    />
    <img
      src={Object {}}
    />
    <img
      src={Object {}}
    />
    <img
      src={Object {}}
    />
    <img
      src={Object {}}
    />
    <img
      src={Object {}}
    />
    <img
      src={Object {}}
    />
    <img
      src={Object {}}
    />
    <img
      src={Object {}}
    />
  </div>
  <h3
    className="book-detail-author"
  >
    Author
  </h3>
  <h2
    className="book-detail-name"
  >
    Name
  </h2>
  <div
    className="book-detail-genres"
  >
    <span
      className="book-detail-genre"
    >
      first
    </span>
    <span
      className="book-detail-genre"
    >
      second
    </span>
  </div>
  <div
    className="book-detail-description"
  >
    description
  </div>
  <div
    className="book-detail-purchase"
  >
    <span
      className="book-detail-price"
    >
      20
      Â£
    </span>
    <span>
      In cart: 
      0
    </span>
    <button
      className="add-book-button book-detail-purchase-button"
      onClick={[Function]}
    >
      add
    </button>
    <button
      className="remove-book-button book-detail-purchase-button"
      disabled={true}
      onClick={[Function]}
    >
      remove
    </button>
  </div>
</div>
`;
